File: apps/mobile/lib/core/constants/session_mode.dart
typedef SessionMode = String;

const sessionModes = ['agent', 'plan', 'ask'];

bool isSessionMode(String? value) {
  return value != null && sessionModes.contains(value);
}

SessionMode resolveSessionMode({String? sessionMode}) {
  return isSessionMode(sessionMode) ? sessionMode! : 'agent';
}

SessionMode normalizeSessionMode(String? value) {
  return isSessionMode(value) ? value! : 'agent';
}

bool isAskSessionMode({String? sessionMode}) =>
    resolveSessionMode(sessionMode: sessionMode) == 'ask';

bool isPlanSessionMode({String? sessionMode}) =>
    resolveSessionMode(sessionMode: sessionMode) == 'plan';

File: apps/desktop/src/shared/session-mode.ts
export const SESSION_MODES = ["agent", "plan", "ask"] as const;

export type SessionMode = (typeof SESSION_MODES)[number];

export interface SessionModeSource {
  sessionMode?: SessionMode;
}

export function isSessionMode(value: unknown): value is SessionMode {
  return typeof value === "string" && (SESSION_MODES as readonly string[]).includes(value);
}

export function normalizeSessionMode(value: unknown): SessionMode {
  return isSessionMode(value) ? value : "agent";
}

export function resolveSessionMode(source?: SessionModeSource): SessionMode {
  return normalizeSessionMode(source?.sessionMode);
}

export function isAskSessionMode(source?: SessionModeSource): boolean {
  return resolveSessionMode(source) === "ask";
}

export function isPlanSessionMode(source?: SessionModeSource): boolean {
  return resolveSessionMode(source) === "plan";
}

/**
 * Composer plus-menu Plan/Ask rows toggle: clicking the already-active mode
 * exits to agent. Switching between plan and ask still selects the clicked mode.
 */
export function togglePlusMenuSessionMode(
  current: SessionMode,
  action: Extract<SessionMode, "plan" | "ask">,
): SessionMode {
  return current === action ? "agent" : action;
}
